internal/controller/certificates/policies/checks.go

```diff
@@ -154,19 +159,58 @@ func SecretKeystoreFormatMatchesSpec(input Input) (string, string, bool) {
 	return "", "", false
 }
 
-func SecretIssuerAnnotationsNotUpToDate(input Input) (string, string, bool) {
-	name := input.Secret.Annotations[cmapi.IssuerNameAnnotationKey]
-	kind := input.Secret.Annotations[cmapi.IssuerKindAnnotationKey]
-	group := input.Secret.Annotations[cmapi.IssuerGroupAnnotationKey]
-	if name != input.Certificate.Spec.IssuerRef.Name ||
+// SecretCertificateNameAnnotationMismatch - When the certificate name annotation is
+// defined, it must match the certificate name.
+func SecretCertificateNameAnnotationMismatch(input Input) (string, string, bool) {
+	name, ok := input.Secret.Annotations[cmapi.CertificateNameKey]
+	if ok && // only check if annotation is present
+		name != input.Certificate.Name {
+		return IncorrectCertificate, fmt.Sprintf("Issuing certificate as Secret was previously issued for Certificate %q", name), true
+	}
+	return "", "", false
+}
+
+// SecretIssuerAnnotationsMismatch - When the issuer annotations are defined,
+// it must match the issuer ref.
+func SecretIssuerAnnotationsMismatch(input Input) (string, string, bool) {
+	name, ok1 := input.Secret.Annotations[cmapi.IssuerNameAnnotationKey]
+	kind, ok2 := input.Secret.Annotations[cmapi.IssuerKindAnnotationKey]
+	group, ok3 := input.Secret.Annotations[cmapi.IssuerGroupAnnotationKey]
+	if (ok1 || ok2 || ok3) && // only check if an annotation is present
+		name != input.Certificate.Spec.IssuerRef.Name ||
 		!issuerKindsEqual(kind, input.Certificate.Spec.IssuerRef.Kind) ||
 		!issuerGroupsEqual(group, input.Certificate.Spec.IssuerRef.Group) {
-		return IncorrectIssuer, fmt.Sprintf("Issuing certificate as Secret was previously issued by %s", formatIssuerRef(name, kind, group)), true
+		return IncorrectIssuer, fmt.Sprintf("Issuing certificate as Secret was previously issued by %q", formatIssuerRef(name, kind, group)), true
+	}
+	return "", "", false
+}
+
+func SecretPublicKeysDiffersFromCurrentCertificateRequest(input Input) (string, string, bool) {
```

Comment:
This is a very important check that was missing previously: if the CSR blob is not signed by the private key in the secret, we know that the CertificateRequest resource is not a good representation of what is stored in the Secret resource. If the keys match, we are a bit more certain that the Secret matches the CurrentRevisionRequest.

Reply:
I agree that this is a useful new check. Please put this in a separate PR, with unit-tests and function documentation for this new function.

Reply:
@wallrj see https://github.com/cert-manager/cert-manager/pull/6168